File: src3/Advisor.java
import java.util.ArrayList;

public class Advisor extends Lecturer {

    // Attributes
    private ArrayList<Student> studentList = new ArrayList<>();
    private ArrayList<Student> awaitingStudents = new ArrayList<>();
    private Student advStudent;
    private ArrayList<String> unreadNotifications = new ArrayList<>();
    private ArrayList<CourseSection> approvedCourses = new ArrayList<>();
    private ArrayList<Notification> notifications = new ArrayList<>();
    private ArrayList<String> readNotifications = new ArrayList<>();

    // Constructor
    public Advisor(ID lecturerID, String firstName, String lastName, Password password, ArrayList<Student> studentList, ArrayList<Student> awaitingStudents) {
        super(lecturerID, firstName, lastName);
        this.setStudentList(studentList);
        this.setAwaitingStudents(awaitingStudents != null ? awaitingStudents : new ArrayList<>());
    }

    public Advisor(ID lecturerID) {
        super(lecturerID);
    }

    // Add a course to the approved list
    public void addToApprovedCourses(CourseSection courseSection) {
        if (courseSection != null) {
            this.approvedCourses.add(courseSection);
        }
    }

    // Display all students assigned to this advisor
    public void viewStudentList() {
        if (studentList.isEmpty()) {
            System.out.println("No students assigned to this advisor.");
            return;
        }

        for (Student student : studentList) {
            System.out.println("ID: " + student.getStudentID().getId());
            System.out.println("First Name: " + student.getFirstName());
            System.out.println("Last Name: " + student.getLastName() + "\n");
        }
    }

    // Select the next student from the awaiting list
    public void selectNextStudent(int index) {
        if (index < 1 || index > awaitingStudents.size()) {
            System.out.println("Invalid index. Please select a valid student.");
            return;
        }
        this.setAdvStudent(awaitingStudents.get(index - 1));
    }

    // Find students who are awaiting approval
    public void findAwaitingStudents() {
        for (Student student : studentList) {
            if (!awaitingStudents.contains(student) && student.getApproveRequest() == 1) {
                awaitingStudents.add(student);
            }
        }
    }

    // Remove a student from the awaiting list
    private void removeAwaitingStudent(Student student) {
        if (student != null) {
            awaitingStudents.remove(student);
        }
    }

    // Approve the courses selected by the advisor's current student
    public void approveCourses() {
        if (advStudent == null) {
            System.out.println("No student selected for approval.");
            return;
        }

        ArrayList<CourseSection> chosenCourses = advStudent.getChosenCourses();
        if (chosenCourses == null || chosenCourses.isEmpty()) {
            System.out.println("No courses to approve for the selected student.");
            return;
        }

        for (CourseSection courseSection : chosenCourses) {
            advStudent.addToApprovedCourses(courseSection);
            courseSection.enrollStudent(advStudent);
        }

        finalizeApproval(2);
    }

    // Disapprove the courses selected by the advisor's current student
    public void disapproveCourses() {
        if (advStudent == null) {
            System.out.println("No student selected for disapproval.");
            return;
        }

        ArrayList<CourseSection> chosenCourses = advStudent.getChosenCourses();
        if (chosenCourses != null) {
            for (CourseSection courseSection : chosenCourses) {
                courseSection.setCapacity(courseSection.getCapacity() + 1);
            }
        }

        finalizeApproval(3);
    }

    // Finalize the course approval process
    private void finalizeApproval(int approvalStatus) {
        advStudent.setChosenCourses(null);
        advStudent.setApproveRequest(approvalStatus);
        removeAwaitingStudent(advStudent);
        sendNotification("", approvalStatus);
        setAdvStudent(null);
    }

    // Add a notification to the advisor
    public void addNotification(Notification message) {
        if (message != null) {
            notifications.add(message);
        }
    }

    // Send a notification to the current student
    public void sendNotification(String message, int req) {
        if (advStudent == null) {
            System.out.println("No student selected for sending notifications.");
            return;
        }

        String sendMessage = message.isEmpty() 
            ? (req == 2 ? "The request is approved!" : "The request is disapproved!") 
            : message;

        advStudent.addNotification(sendMessage);
    }

    // Mark all notifications as read
    public void notificationsSeen() {
        readNotifications.addAll(unreadNotifications);
        unreadNotifications.clear();
    }

    // Getters and Setters
    public ArrayList<Student> getStudentList() {
        return studentList;
    }

    public void setStudentList(ArrayList<Student> studentList) {
        this.studentList = studentList != null ? studentList : new ArrayList<>();
    }

    public ArrayList<Student> getAwaitingStudents() {
        return awaitingStudents;
    }

    public void setAwaitingStudents(ArrayList<Student> awaitingStudents) {
        this.awaitingStudents = awaitingStudents != null ? awaitingStudents : new ArrayList<>();
    }

    public Student getAdvStudent() {
        return advStudent;
    }

    public void setAdvStudent(Student advStudent) {
        this.advStudent = advStudent;
    }

    public ArrayList<String> getUnreadNotifications() {
        return unreadNotifications;
    }

    public void setUnreadNotifications(ArrayList<String> unreadNotifications) {
        this.unreadNotifications = unreadNotifications;
    }

    public ArrayList<String> getReadNotifications() {
        return readNotifications;
    }

    public void setReadNotifications(ArrayList<String> readNotifications) {
        this.readNotifications = readNotifications;
    }

    public ArrayList<Notification> getNotifications() {
        return notifications;
    }

    public void setNotifications(ArrayList<Notification> notifications) {
        this.notifications = notifications != null ? notifications : new ArrayList<>();
    }
}

File: Iteration3/Python/Advisor.py
class Advisor(Lecturer):
    
    def __init__(self, lecturer_id, first_name=None, last_name=None, password=None, student_list=None, awaiting_students=None):
        super().__init__(lecturer_id, first_name, last_name)
        self.student_list = student_list if student_list else []
        self.awaiting_students = awaiting_students if awaiting_students else []
        self.adv_student = None
        self.unread_notifications = []
        self.approved_courses = []
        self.notifications = []
        self.read_notifications = []

    def add_to_approved_courses(self, course_section):
        self.approved_courses.append(course_section)

    def view_student_list(self):
        for student in self.student_list:
            print(f"ID: {student.get_student_id().get_id()}")
            print(f"First Name: {student.get_first_name()}")
            print(f"Last Name: {student.get_last_name()}\n")

    def select_next_student(self, index):
        self.adv_student = self.awaiting_students[index - 1]

    def find_awaiting_students(self):
        for student in self.student_list:
            if student not in self.awaiting_students and student.get_approve_request() == 1:
                self.awaiting_students.append(student)

    def remove_awaiting_student(self, student):
        self.awaiting_students.remove(student)

    def approve_courses(self):
        selected_courses = self.adv_student.get_chosen_courses()
        for course_section in selected_courses:
            self.adv_student.add_to_approved_courses(course_section)
            course_section.enroll_student(self.adv_student)
        self.adv_student.set_chosen_courses(None)
        self.adv_student.set_approve_request(2)
        self.remove_awaiting_student(self.adv_student)
        self.adv_student = None

    def disapprove_courses(self):
        selected_courses = self.adv_student.get_chosen_courses()
        for course_section in selected_courses:
            course_section.set_capacity(course_section.get_capacity() + 1)
        self.adv_student.set_chosen_courses(None)
        self.adv_student.set_approve_request(3)
        self.remove_awaiting_student(self.adv_student)
        self.adv_student = None

    def add_notification(self, notification):
        self.notifications.append(notification)

    def send_notification(self, message, req):
        if not message:
            if req == 2:
                message = "The request is approved!"
            elif req == 3:
                message = "The request is disapproved!"
        self.adv_student.add_notification(message)

    def add_unread_notification(self, message):
        self.unread_notifications.append(message)

    def notifications_seen(self):
        self.read_notifications.extend(self.unread_notifications)
        self.unread_notifications.clear()

    def get_student_list(self):
        return self.student_list

    def set_student_list(self, student_list):
        self.student_list = student_list

    def get_awaiting_students(self):
        return self.awaiting_students

    def set_awaiting_students(self, awaiting_students):
        self.awaiting_students = awaiting_students

    def get_adv_student(self):
        return self.adv_student

    def set_adv_student(self, adv_student):
        self.adv_student = adv_student

    def get_given_courses(self):
        return super().get_given_courses()

    def get_unread_notifications(self):
        return self.unread_notifications

    def set_unread_notifications(self, unread_notifications):
        self.unread_notifications = unread_notifications

    def get_read_notifications(self):
        return self.read_notifications

    def set_read_notifications(self, read_notifications):
        self.read_notifications = read_notifications

    def get_notifications(self):
        return self.notifications

    def set_notifications(self, notifications):
        self.notifications = notifications